### ISSUE TITLE
Widget navigation steals editor focus

### ISSUE THREAD

**Issue by rebornix** [microsoft/vscode#200259]
Does this issue occur when all extensions are disabled?: Yes/No

```
2023-12-07 09:49:34.783 [info] [KeybindingService]: \ From 22 keybinding entries, matched widgetNavigation.focusNext, when: inputFocus && navigableContainerFocused || navigableContainerFocused && !listFocus || navigableContainerFocused && listScrollAtBoundary == 'both' || navigableContainerFocused && listScrollAtBoundary == 'bottom', source: built-in.
2023-12-07 09:49:34.784 [info] [KeybindingService]: / Received  keydown event - modifiers: [meta], code: ArrowDown, keyCode: 40, key: ArrowDown
2023-12-07 09:49:34.784 [info] [KeybindingService]: | Converted keydown event - modifiers: [meta], code: ArrowDown, keyCode: 18 ('DownArrow')
2023-12-07 09:49:34.784 [info] [KeybindingService]: | Resolving meta+[ArrowDown]
2023-12-07 09:49:34.784 [info] [KeybindingService]: \ From 22 keybinding entries, matched widgetNavigation.focusNext, when: inputFocus && navigableContainerFocused || navigableContainerFocused && !listFocus || navigableContainerFocused && listScrollAtBoundary == 'both' || navigableContainerFocused && listScrollAtBoundary == 'bottom', source: built-in.
2023-12-07 09:49:34.785 [info] [KeybindingService]: + Invoking command widgetNavigation.focusNext.
```

Not sure how did I end with this state but often I can't `cmd+down`anymore in editor and the shortcut is stolen by widgetNavigation

**Comment by meganrogge** [comment]
@hsfzxjy would be awesome if you could look into this 🙏🏼

**Comment by meganrogge** [comment]
@rebornix any steps to reproduce would be appreciated

**Comment by hsfzxjy** [comment]
@rebornix I will investigate this. Could you provide some reproduction steps or observation on the trigger condition?

**Comment by AlekSi** [comment]
I noticed that bug about a month ago, and it has been driving me crazy since. Finally, I found that issue.

I'm not sure how to reliably to reproduce it, though. Everything works for some time after restart, but then ⌘+up/down stops working.

**Comment by hsfzxjy** [comment]
@AlekSi Hi. It'd be great if you could share some information regarding the following questions:

1. What's your desired behavior of pressing cmd+up/down in editor? Is it for scrolling lines? Sorry if this may seem dumb but I have seldom used this shortcut. The answer would help me troubleshooting.
2. Did you remember comments/chat or some other panels being opened in the editor when the bug shown up?
3. Where did the focus go after this bug happened?

**Comment by meganrogge** [comment]
From @rebornix , still no repro steps, but:

```
2023-12-18 12:06:32.748 [info] [KeybindingService]: \ From 22 keybinding entries, matched widgetNavigation.focusNext, when: inputFocus && navigableContainerFocused || navigableContainerFocused && !listFocus || navigableContainerFocused && listScrollAtBoundary == 'both' || navigableContainerFocused && listScrollAtBoundary == 'bottom', source: built-in.
2023-12-18 12:06:32.748 [info] [KeybindingService]: / Received  keydown event - modifiers: [meta], code: ArrowDown, keyCode: 40, key: ArrowDown
2023-12-18 12:06:32.748 [info] [KeybindingService]: | Converted keydown event - modifiers: [meta], code: ArrowDown, keyCode: 18 ('DownArrow')
2023-12-18 12:06:32.749 [info] [KeybindingService]: | Resolving meta+[ArrowDown]
2023-12-18 12:06:32.749 [info] [KeybindingService]: \ From 22 keybinding entries, matched widgetNavigation.focusNext, when: inputFocus && navigableContainerFocused || navigableContainerFocused && !listFocus || navigableContainerFocused && listScrollAtBoundary == 'both' || navigableContainerFocused && listScrollAtBoundary == 'bottom', source: built-in.
2023-12-18 12:06:32.749 [info] [KeybindingService]: + Invoking command widgetNavigation.focusNext.
2023-12-18 12:06:33.046 [info] [KeybindingService]: + Storing single modifier for possible chord meta.
2023-12-18 12:06:33.349 [info] [KeybindingService]: + Clearing single modifier due to 300ms elapsed.

```

**Comment by rebornix** [comment]
@hsfzxjy sorry that I didn't have a reproduce and I didn't even realize that I already reported it when creating a second issue today.

**Comment by rebornix** [comment]
I inspected the context keys, here is what I found, while the focus is in the editor

```
inputFocus: true // text editor input is focused
navigableContainerFocused: true
```

which matches the when clause

```
inputFocus  && navigableContainerFocused 
|| navigableContainerFocused && !listFocus 
|| navigableContainerFocused && listScrollAtBoundary == 'both' 
|| navigableContainerFocused && listScrollAtBoundary == 'bottom'
```

`navigableContainerFocused: true` seems an oversight to me.

**Comment by hsfzxjy** [comment]
@rebornix Sounds like a broken state management of this feature.

Can you remember whether or not you had focus in any of these components before the bug happened? :

 + Settings Editor
 + Keyboard Shortcut Editor
 + Debug Console Panel
 + Problems Panel
 + Extensions Panel
 + Testing Panel
 + Comments Panel
 + The in-editor comment thread widget

Also, when the bug happened, you might check whether switching focus to these components and then back to the editor would recover from the problem.

**Comment by AlekSi** [comment]
> What's your desired behavior of pressing cmd+up/down in editor? Is it for scrolling lines?

⌘+up is mapped to multiple commands by default; the relevant seems to be `cursorTop.`

---

For now, I just disabled `widgetNavigation.focusPrevious`:

```json
	{
		"key": "cmd+up",
		"command": "-widgetNavigation.focusPrevious",
		"when": "inputFocus && navigableContainerFocused || navigableContainerFocused && !listFocus || navigableContainerFocused && listScrollAtBoundary == 'both' || navigableContainerFocused && listScrollAtBoundary == 'top'"
	}
```

**Comment by rebornix** [comment]
It happened really randomly and I have no clue how it's triggered. Just ran into this again ;(

**Comment by rebornix** [comment]
Adding to this milestone for now as I'm afraid that more users might get to it after Jan release.

**Comment by meganrogge** [comment]
@hsfzxjy , @rebornix is often navigating between Jupyter notebook cells, that might be key to reproduce this

**Comment by hsfzxjy** [comment]
Still unable to reproduce. Is it possible that we make a special build for the affected users, where I will add some logging for debug purpose?

**Comment by samgermain** [comment]
> ```json
> {
> 	"key": "cmd+up",
> 	"command": "-widgetNavigation.focusPrevious",
> 	"when": "inputFocus && navigableContainerFocused || navigableContainerFocused && !listFocus || navigableContainerFocused && listScrollAtBoundary == 'both' || navigableContainerFocused && listScrollAtBoundary == 'top'"
> }
> ```

Where do you add this?

**Comment by samgermain** [comment]
# Fix

## Step 1

![](https://i.imgur.com/BgoPWkt.png)

## Step 2

![](https://i.imgur.com/Db5KAaa.png)

**Comment by rebornix** [comment]
I ran into this again, and it's not the first time for the last month ;(

I can anticipate how challenging it is to troubleshoot and fix this problem considering it's flaky, but we need to do something. Here is my suggestion:

* Add a setting (can be hidden) to track all focus change relevant to `navigableContainerFocused`.
* Add a troubleshoot command and when this bug reproduces, run the command to get all the logs.

@meganrogge I would recommend we have it on the iteration plan if we couldn't spare time on it.

**Comment by meganrogge** [comment]
@hsfzxjy would you be up for creating a PR that accomplishes the above?

**Comment by hsfzxjy** [comment]
Would be glad to. But I am not quite familiar with the logging part. Could you enlight some instructions to log under such a case?

**Comment by meganrogge** [comment]
Yes, I'd suggest using the `ILogService` to log all focus changes. I think you'll just want to use log level of `trace` when the setting mentioned above is off and use log level of `debug` when it's enabled. 

I don't think we need a command to get all of the logs. Users who repro this frequently can just enable that setting IMO.

**Comment by meganrogge** [comment]
@rebornix will the above work for you or can you further explain what you mean here?

> run the command to get all the logs.

**Comment by hsfzxjy** [comment]
@meganrogge Hi. I've tried using ILogService. The output of the logging seems to appear in `Output - Window` and mixed with logging messages from other part of VSCode. Is it what we expect or should I create a separate logging channel for this?

**Comment by meganrogge** [comment]
I think logging to `Output - Window` is fine because we'll only do this when the setting is enabled. Thanks @hsfzxjy

**Comment by rebornix** [comment]
@meganrogge using log service and show them in output channel makes good sense to me

**Comment by hsfzxjy** [comment]
@meganrogge @rebornix Hi. I've implemented the function in #209357 . Please check if this is desired 🙏🏼

**Comment by meganrogge** [comment]
@rebornix , this will be in shortly. Could you pls add `workbench.navigibleContainer.enableDebug` to your settings and set log level for the `window` to `debug`?

**Comment by rebornix** [comment]
Thanks, updated the setting and log level and will see what I can get when I reproduce again.
